File: Packages/StrandDesign/Sources/StrandDesign/RingDial.swift
import SwiftUI

// MARK: - RingDial (WHOOP score dial)
//
// The WHOOP score ring: a thick, rounded-cap arc on a faint dark-grey track, starting
// at 12 o'clock and filling clockwise to `value` (0...1). A D-DIN numeral sits centred,
// with a tracked ALL-CAPS Montserrat label. The arc draws in from empty over 0.8s
// ease-out on appear (snaps instantly under Reduce Motion). No bloom, WHOOP-flat.
//
// Three sizes match the reference frames:
//   trio  96pt, the home three-up row (Rest / Charge / Effort). Label sits BELOW the ring.
//   hero  260pt, the detail-screen headline dial. Numeral about 72pt, label INSIDE the ring.
//   mini  44pt, small card/tile dials. Numeral only.
//
// Replaces the LiquidVessel gauge as the app's primary score glyph. Public API is stable:
// screens construct it by (value, display, label, tint, size).

public struct RingDial: View {

    /// The three canonical dial sizes.
    public enum Size {
        case trio   // 96pt, home three-up
        case hero   // 260pt, detail hero
        case mini   // 44pt, card/tile

        /// Outer diameter of the ring.
        var diameter: CGFloat {
            switch self {
            case .trio: return 96
            case .hero: return 260
            case .mini: return 44
            }
        }

        /// Stroke thickness: a thick arc at about 10% of the diameter (WHOOP-faithful).
        var lineWidth: CGFloat { diameter * 0.10 }

        /// Centre numeral point size (D-DIN). Hero is the ~72pt headline score.
        var numberSize: CGFloat {
            switch self {
            case .trio: return 26
            case .hero: return 72
            case .mini: return 15
            }
        }

        /// Caps-label point size (Montserrat semibold).
        var labelSize: CGFloat {
            switch self {
            case .trio: return 10
            case .hero: return 13
            case .mini: return 8
            }
        }

        /// Tracking for the caps label (about 0.12em).
        var labelTracking: CGFloat { labelSize * 0.12 }

        /// The hero carries its label INSIDE the ring beneath the numeral; the smaller
        /// dials place it below the ring so the centre stays uncrowded.
        var labelInsideRing: Bool { self == .hero }

        /// The mini dial shows the numeral only (label reserved for accessibility).
        var showsLabel: Bool { self != .mini }
    }

    /// Fill fraction 0...1 (e.g. recoveryScore / 100, or dayStrain / 21).
    public var value: Double
    /// The centred read-out string, e.g. "85%" or "14.2".
    public var display: String
    /// The tracked ALL-CAPS label, e.g. "RECOVERY" / "DAY STRAIN".
    public var label: String
    /// Arc tint: a band/domain colour from `StrandPalette`.
    public var tint: Color
    /// Which of the three canonical sizes to render.
    public var size: Size

    public init(value: Double, display: String, label: String, tint: Color, size: Size) {
        self.value = value
        self.display = display
        self.label = label
        self.tint = tint
        self.size = size
    }

    @State private var animatedFraction: Double = 0
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    private var fraction: Double { min(max(value, 0), 1) }

    public var body: some View {
        Group {
            if size.showsLabel && !size.labelInsideRing {
                VStack(spacing: 8) {
                    ring
                    capsLabel
                }
            } else {
                ring
            }
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(Text(label))
        .accessibilityValue(Text(display))
        .onAppear {
            if reduceMotion {
                animatedFraction = fraction
            } else {
                withAnimation(.easeOut(duration: 0.8)) { animatedFraction = fraction }
            }
        }
        .onChangeCompat(of: fraction) { newValue in
            if reduceMotion {
                animatedFraction = newValue
            } else {
                withAnimation(.easeOut(duration: 0.8)) { animatedFraction = newValue }
            }
        }
    }

    // MARK: Ring + centre content

    private var ring: some View {
        ZStack {
            // Faint dark-grey full-circle track (hairline #2A343A at 30% opacity).
            RingDialArc(fraction: 1, lineWidth: size.lineWidth)
                .stroke(
                    StrandPalette.hairline.opacity(0.30),
                    style: StrokeStyle(lineWidth: size.lineWidth, lineCap: .round)
                )

            // Filled arc: solid tint, rounded caps, 12 o'clock start, clockwise.
            RingDialArc(fraction: animatedFraction, lineWidth: size.lineWidth)
                .stroke(
                    tint,
                    style: StrokeStyle(lineWidth: size.lineWidth, lineCap: .round)
                )

            // Centre read-out: D-DIN numeral, plus the caps label for the hero.
            VStack(spacing: size == .hero ? 4 : 1) {
                Text(display)
                    .font(StrandFont.number(size.numberSize, weight: .bold))
                    .tracking(-size.numberSize * 0.02)
                    .foregroundStyle(StrandPalette.textPrimary)
                    .minimumScaleFactor(0.6)
                    .lineLimit(1)
                if size.labelInsideRing {
                    capsLabel
                }
            }
            .padding(.horizontal, size.lineWidth)
        }
        .frame(width: size.diameter, height: size.diameter)
    }

    private var capsLabel: some View {
        Text(label)
            .font(StrandFont.text(size.labelSize, weight: .semibold))
            .tracking(size.labelTracking)
            .textCase(.uppercase)
            .foregroundStyle(StrandPalette.textSecondary)
            .lineLimit(1)
            .allowsHitTesting(false)
    }
}

// MARK: - Ring arc shape
//
// A closed-track fill arc: starts at 12 o'clock and sweeps CLOCKWISE by `fraction`
// of the full circle. `clockwise: false` in `addArc` draws clockwise on screen because
// SwiftUI's y-axis points down (same convention as `RecoveryArc`).

/// The RingDial fill arc: 12 o'clock start, clockwise, `fraction` of 360 degrees.
public struct RingDialArc: Shape {
    public var fraction: Double
    public var lineWidth: CGFloat

    public init(fraction: Double, lineWidth: CGFloat) {
        self.fraction = fraction
        self.lineWidth = lineWidth
    }

    public var animatableData: Double {
        get { fraction }
        set { fraction = newValue }
    }

    public func path(in rect: CGRect) -> Path {
        let radius = (min(rect.width, rect.height) - lineWidth) / 2
        let center = CGPoint(x: rect.midX, y: rect.midY)
        let start = Angle.degrees(-90)   // 12 o'clock
        let end = Angle.degrees(-90 + 360 * min(max(fraction, 0), 1))
        var path = Path()
        path.addArc(
            center: center,
            radius: radius,
            startAngle: start,
            endAngle: end,
            clockwise: false
        )
        return path
    }
}

#if DEBUG
#Preview("RingDial sizes") {
    VStack(spacing: 40) {
        HStack(spacing: 24) {
            RingDial(value: 0.80, display: "80%", label: "Rest",
                     tint: StrandPalette.restColor, size: .trio)
            RingDial(value: 0.85, display: "85%", label: "Charge",
                     tint: StrandPalette.recoveryBand(85), size: .trio)
            RingDial(value: 0.68, display: "14.2", label: "Effort",
                     tint: StrandPalette.strainColor(68), size: .trio)
        }
        RingDial(value: 0.85, display: "85%", label: "Recovery",
                 tint: StrandPalette.recoveryBand(85), size: .hero)
        HStack(spacing: 16) {
            RingDial(value: 0.55, display: "55", label: "Mini",
                     tint: StrandPalette.recoveryBand(55), size: .mini)
            RingDial(value: 0.30, display: "30", label: "Mini",
                     tint: StrandPalette.recoveryBand(30), size: .mini)
        }
    }
    .padding(40)
    .frame(width: 520, height: 640)
    .background(StrandPalette.canvasGradient)
    .preferredColorScheme(.dark)
}
#endif

File: android/app/src/main/java/com/noop/ui/components/RingDial.kt
package com.noop.ui.components

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawWithCache
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.noop.ui.Motion
import com.noop.ui.NoopType
import com.noop.ui.Palette
import com.noop.ui.rememberReduceMotion

// MARK: - RingDial (the WHOOP score ring, Compose parity with StrandDesign/RingDial.swift)
//
// The single WHOOP-faithful ring that REPLACES the old LiquidVessel gauge across the app. It mirrors the
// Swift `RingDial` contract exactly:
//
//   public struct RingDial: View {
//       public enum Size { case trio, hero, mini }   // 96pt, 260pt, 44pt
//       let value: Double        // 0...1 fill fraction
//       let display: String      // "85%" / "14.2"
//       let label: String        // "RECOVERY" tracked caps
//       let tint: Color          // band colour from palette
//       let size: Size
//   }
//
// Anatomy (matches the Swift view and the reference screenshots):
//   1. a thick ROUNDED-CAP arc, stroke about 10% of the diameter,
//   2. sitting on a full-circle TRACK at Palette.hairline (#2A343A) drawn at 30% opacity (a hairline well),
//   3. starting at 12 o'clock and sweeping CLOCKWISE,
//   4. animating its fill 0 to [value] on first composition over 0.8s ease-out (honouring Reduce Motion),
//   5. a centred D-DIN numeral ([display], hero about 72sp) with the tracked-caps [label] below it in
//      Montserrat SemiBold (the WHOOP two-font rule: numerals D-DIN, words Montserrat).
//
// Colour comes ONLY from palette tokens: [tint] is a band/domain colour the caller samples from
// Palette (e.g. Palette.recoveryBand(score) or Palette.effortColor); the track + text are palette tokens.

/** The three ring sizes, matching the Swift `RingDial.Size` (dp == the Swift pt values). */
enum class RingDialSize(val diameter: Dp) {
    /** Home-screen trio (Rest / Charge / Effort). 96dp. */
    Trio(96.dp),

    /** Detail-screen hero. 260dp. */
    Hero(260.dp),

    /** Mini card accent. 44dp. */
    Mini(44.dp),
}

/**
 * A WHOOP score ring. [value] is the 0..1 fill fraction; [display] is the centred numeral text
 * ("85%" / "14.2"); [label] is the tracked-caps word below it ("RECOVERY"); [tint] is the arc's band
 * colour (a Palette token the caller samples). [size] picks the trio / hero / mini geometry.
 *
 * The arc starts at 12 o'clock, sweeps clockwise, and animates 0 to [value] on first appearance over 0.8s
 * ease-out (snaps instantly under Reduce Motion). Mirrors the iOS `RingDial` view.
 */
@Composable
fun RingDial(
    value: Double,
    display: String,
    label: String,
    tint: Color,
    size: RingDialSize,
    modifier: Modifier = Modifier,
    showsLabel: Boolean = true,
) {
    val diameter = size.diameter
    val target = value.toFloat().coerceIn(0f, 1f)
    val reduced = rememberReduceMotion()

    // Fill animates 0 to value on FIRST composition (WHOOP's draw-in), over 0.8s ease-out. Under Reduce
    // Motion it lands immediately (tween 0). Mirrors iOS `.animation(.easeOut(duration: 0.8))` on appear.
    var started by remember { mutableStateOf(false) }
    LaunchedEffect(Unit) { started = true }
    val animatedFraction by animateFloatAsState(
        targetValue = if (started) target else 0f,
        animationSpec = if (reduced) tween(0) else tween(durationMillis = 800, easing = Motion.easeOut),
        label = "ringDialFill",
    )

    // Track colour: the WHOOP dark-grey hairline well at 30% opacity (design contract "#2A343A at 30%").
    val trackColor = Palette.hairline.copy(alpha = 0.30f)

    Box(
        modifier = modifier
            .size(diameter)
            .semantics { contentDescription = if (showsLabel) "$label $display" else display },
        contentAlignment = Alignment.Center,
    ) {
        Box(
            modifier = Modifier
                .size(diameter)
                .drawWithCache {
                    // Stroke about 10% of the diameter (thick WHOOP arc), round caps on both track and fill.
                    val stroke = size.minDimension * 0.10f
                    val inset = stroke / 2f
                    val d = size.minDimension
                    val arcSize = Size(d - stroke, d - stroke)
                    val topLeft = Offset(
                        (size.width - d) / 2f + inset,
                        (size.height - d) / 2f + inset,
                    )
                    val capStroke = Stroke(width = stroke, cap = StrokeCap.Round)
                    onDrawBehind {
                        // Full-circle track (the carved well the arc sits in), static, so the fraction-driven
                        // arc below is the only per-frame draw.
                        drawArc(
                            color = trackColor,
                            startAngle = 0f,
                            sweepAngle = 360f,
                            useCenter = false,
                            topLeft = topLeft,
                            size = arcSize,
                            style = capStroke,
                        )
                        // The fill arc, from 12 o'clock (-90 degrees) clockwise. Only drawn once there is
                        // actual progress: a near-zero round-capped arc otherwise renders as a stray dot at 12
                        // o'clock on Android's Canvas (the empty state reads as the clean track alone).
                        val sweep = animatedFraction.coerceIn(0f, 1f) * 360f
                        if (animatedFraction > 0.001f) {
                            drawArc(
                                color = tint,
                                startAngle = -90f,
                                sweepAngle = sweep,
                                useCenter = false,
                                topLeft = topLeft,
                                size = arcSize,
                                style = capStroke,
                            )
                        }
                    }
                },
        )

        // Centred numeral (D-DIN) + tracked caps label (Montserrat SemiBold), scaled to the ring size.
        val numberSp = when (size) {
            RingDialSize.Hero -> 72f
            RingDialSize.Trio -> 26f
            RingDialSize.Mini -> 15f
        }
        val labelSp = when (size) {
            RingDialSize.Hero -> 13f
            RingDialSize.Trio -> 10f
            RingDialSize.Mini -> 8f
        }
        // The mini ring is too small to carry a word under the numeral; it shows the numeral alone.
        val labelVisible = showsLabel && size != RingDialSize.Mini && label.isNotEmpty()

        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            modifier = Modifier.padding(PaddingValues(horizontal = diameter * 0.12f)),
        ) {
            Text(
                text = display,
                style = NoopType.number(numberSp, FontWeight.Bold),
                color = Palette.textPrimary,
                maxLines = 1,
                textAlign = TextAlign.Center,
            )
            if (labelVisible) {
                Text(
                    text = label.uppercase(),
                    style = NoopType.text(labelSp, FontWeight.SemiBold).copy(letterSpacing = (labelSp * 0.14f).sp),
                    color = Palette.textSecondary,
                    maxLines = 1,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.padding(top = diameter * 0.02f),
                )
            }
        }
    }
}
